### DISCUSSION TITLE
Post Not Working

### DISCUSSION THREAD

**Discussion by edris6** [rwf2/Rocket#2450] [Questions] [answered]
When i run this server i cant do a post request
server:
```
#![feature(proc_macro_hygiene, decl_macro)]

#[macro_use]
extern crate rocket;


#[get("/")]
fn index() -> &'static str {
    "Hello, world!"
}

#[post("/submit", data = "<data>")]
fn submit(data: String) -> String {
    format!("You submitted: {}", data)
}

fn main() {
    
    rocket::ignite()
        .mount("/", routes![index, submit])
        .launch();
}
```

and when i do a post request with js in web browser
it gives this error
![image](https://user-images.githubusercontent.com/78484256/212533525-23f6bc15-a907-4f87-908f-0ff15f00e4fe.png)
with this code

```
fetch('http://localhost:8000/submit', {
  method: 'POST',
  headers: {
    'Content-Type': 'application/json'
  },
  body: JSON.stringify({data: 'example'})
})
.then(response => response.text())
.then(data => console.log(data))

```

**Comment by blackghost1987** [comment, marked as answer]
I think there are two separate issues here.

1. CORS protection is blocking your request in the browser, so the POST request is not sent at all. For security reasons the browser is sending a preflight request (OPTIONS) before the POST, and it doesn't like the servers response. You have two options here:
- run the backend and frontend on the same domain (IP), so it's not considered a cross-origin call by the browser
- add the required CORS headers to your responses. The best way IMHO to do this is to use the Rocket-Cors crate (https://crates.io/crates/rocket_cors)

2. Even if your POST request arrives, the server will probably return a 404, because your input parameter type is not matching the data you're sending from the JS code: You're sending a JSON object as the body, but on the server side you're expecting a simple string.

If you want to use the same JS code, then you should use `data: Json<SomeStruct>` instead of `data: String`, where SomeStruct is:
```
#[derive(Deserialize)]
struct SomeStruct {
  data: String
}
```

OR if you don't need any more parameters, it's much easier to use a simple JSON string:

in the JS: `JSON.stringify('example')`
in Rocket: `data: Json<String>`

**Reply by edris6** [reply]
Is there any way to send html

**Reply by edris6** [reply]
i know this works: #[get("/")]
fn index() -> &'static str {
    "Hello, world!"
} but send like a file

**Reply by open-schnick** [reply]
> i know this works: #[get("/")] fn index() -> &'static str { "Hello, world!" } but send like a file

Look here: https://api.rocket.rs/v0.5-rc/rocket/fs/struct.NamedFile.html
Also for an easy workaround you can add a CORS Fairing like this:
```rust
#[options("/<_path..>")]
fn cors_response_handler(_path: PathBuf) {
    // NOOP
}

#[launch]
fn rocket() -> _ {
    rocket::build()
        .attach(CORS)
        .mount("/", routes![hello, cors_response_handler])
}

struct CORS;

#[rocket::async_trait]
impl Fairing for CORS {
    fn info(&self) -> Info {
        Info {
            name: "Add CORS headers to responses.",
            kind: Kind::Response,
        }
    }

    async fn on_response<'r>(&self, _request: &'r Request<'_>, response: &mut Response<'r>) {
        response.set_header(Header::new("Access-Control-Allow-Origin", "*"));
        response.set_header(Header::new(
            "Access-Control-Allow-Methods",
            "POST, GET, PATCH, OPTIONS",
        ));
        response.set_header(Header::new("Access-Control-Allow-Headers", "*"));
        response.set_header(Header::new("Access-Control-Allow-Credentials", "true"));
    }
}

```

**Reply by edris6** [reply]
but it would actually be good, if i cant send html files

**Reply by edris6** [reply]
how can i do it

**Reply by blackghost1987** [reply]
I'm not sure how this is is related to the original question about POST requests, but anyways: There's multiple ways to serve HTML files from Rocket. The first thing you have to decide is if you want to serve static HTML files (content doesn't change) or dynamic HTML files (content changing based on the request or the state)? If you are trying to serve static files, the question is if it's one file or multiple files in a directory?

Your options:
1. Serving a single static file with NamedFile, as @open-schnick proposed, see: https://api.rocket.rs/v0.5-rc/rocket/fs/struct.NamedFile.html
2. Serving a bunch of static files from a directory with the FileServer feature, see: https://api.rocket.rs/v0.5-rc/rocket/fs/struct.FileServer.html
3. Serving dynamic files with Templates, see: https://rocket.rs/v0.5-rc/guide/responses/#templates

There are working examples for all these in the examples directory, see:
https://github.com/SergioBenitez/Rocket/tree/v0.5-rc/examples/static-files
and
https://github.com/SergioBenitez/Rocket/tree/v0.5-rc/examples/templating

**Reply by edris6** [reply]
thanks, for the help!
